### DISCUSSION TITLE
How do I type ESC, ENTER and other keys?

### DISCUSSION THREAD

**Discussion by HellShuro** [seleniumbase/SeleniumBase#1346] [Q&A] [answered]
How do I type ESC, ENTER and other keys?
I only see the up and down, left and right approach
Using self.send_Keys(Key.ESCAPE) has no effect

**Comment by HellShuro** [comment]
![image](https://user-images.githubusercontent.com/32864369/171369612-c5dace6d-929d-43ac-8804-23c8bc798f54.png)
Or is it only possible in this way?

**Comment by mdmintz** [comment, marked as answer]
For using special keys, you can use:

```python
from selenium.webdriver.common.keys import Keys
...
    ...
    self.send_keys(SELECTOR, Keys.ESCAPE)
```

But for hitting ENTER/RETURN, you can use ``\n``. Example:

```python
self.type("input#password", "secret_sauce\n")
```

**Comment by akunteseterwork** [comment]
@mdmintz But how to press Keys multiple times without specifying element?

Say, i want like this

<img width="756" alt="image" src="https://user-images.githubusercontent.com/94689744/177008137-932bf7fd-e0c2-4346-97a6-8d8d815c0aa9.png">

But it doesn't work. 
I've tried combining with the traditional selenium with actionchains but it just gave me another error, also it opens 2 browsers since i call for driver and self.

**Reply by mdmintz** [reply]
A selector must always be specified to attach the action to. You can use ``html`` or ``body`` for the page when you don't want to use a specific element on the page.

Example:

```python
from selenium.webdriver.common.keys import Keys
self.send_keys("html", Keys.TAB)
```

**Reply by mdmintz** [reply]
Using ``Keys.TAB`` changes the active element. Then to click the active element, use:

```python
self.click_active_element()
```

**Reply by akunteseterwork** [reply]
<img width="875" alt="image" src="https://user-images.githubusercontent.com/94689744/177009469-7b0bc9d0-9ff6-4f84-90ca-28102deec5e5.png">

YOU'RE A GENIOUS!
Thank you so much, hope you have a great day (y)

**Reply by wackydude** [reply]
Is there a self.send_keys_active_element("whatevs")? Or maybe make send_keys have a default selector of active_element? so if it gets 2 strings, first one is selector and second string is text to type but if only 1 string is provided, assume selector is active_element and provided string is text to type?

**Reply by mdmintz** [reply]
@wackydude: `self.send_keys(self.get_active_element_css(), "TEXT")`

`self.get_active_element_css()` was added this week, so make sure you have the latest SeleniumBase first.

**Comment by ksingha161** [comment]
hey @mdmintz how can we press keys in `sb.cdp`?

**Reply by mdmintz** [reply]
Depends on which keys. Most keys can be pressed with `sb.cdp.type(selector, text)`. There are some working escape sequences such as `\n` or `\r` for Enter and `\b` for Backspace. For other keys like ESC, you may need `PyAutoGUI`.

**Reply by ksingha161** [reply]
I tried `sb.cdp.press_keys("//input[@id='gn-search-input'] \n", text)` but it does not seem to work

**Reply by mdmintz** [reply]
The "\n" should go with the text, not the selector. And you should be using CSS selectors with CDP Mode (not XPath).

**Reply by ksingha161** [reply]
Tried 
``germain = "Paris saint germain\n" 
sb.cdp.press_keys("input#gn-search-input", germain)``

but doesn't seem to work. Pardon me if i'm missing something, i'm new to this amazing project. just need to explore it fully

**Reply by mdmintz** [reply]
If "\n" didn't work, and "\r" didn't work, then use the PyAutoGUI methods for that, eg: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_antibot_login.py
